README.md
This software is still under development. 

The server development is almost finished and can been tested.

The client development is not started.

## Product presentation

MnemoPwd is a client-server application to store and retrieve secret informations.
It is based on OpenSSL cryptographic library and its installation is required both
on the client and the server.

MnemoPwd use the PyElliptic module developped by Yann GUIBET under BSD License,
that is a high level wrapper of OpenSSL (see https://github.com/yann2192/pyelliptic).

See [OpenSSL] (https://www.openssl.org "OpenSSL Homepage") web site for more information.

Require Python 3.4.4 or newer, OpenSSL, GNU/Linux or Mac OS X

Always use the last version for security.

Copyright (c) 2015-2016, Thierry Lemeunier <thierry at lemeunier dot net> under 
BSD Licence

## Features

- One central server serves multiple clients
- Multiple clients can be used on a same node (each with a different user)
- Double level of connection security : the secure SSL/TLS protocol and the secure application protocol
- Server identity can be controlled by X.509 certificat (optional TLS feature)
- Client and server share a master secret via ECDH scheme
- Each secret information is encrypted with the ECIES scheme
- Integrity is controlled by two hmac : a 256-bits hmac for each secret information (ECIES scheme) and a 512-bits hmac for each block
- Possibility of three stages of data encryption by client's configuration
- Keys are computed on the fly by the client and the server
- Keys are never stored by the server or by the client
- Do some self-controls before starting
- Configuration by file and/or by command line options
- Server running in background (only on POSIX systems)
- UML model of the secure application protocol

## Usage

### Start the server

   `./serverctl -h`          --> get a help screen

   `./serverctl --start`     --> start the server

   `./serverctl [--status]`  --> get a status message

   `./serverctl --stop`      --> stop the server

### Start a client

   Client not yet developed :disappointed:

### Certificat usage

To autenticate server, a X.509 certificat can be used. You can use an existing certificat or use
a new self-signed certificat created with OpenSSL. In the last case, please follow next steps:

1. Generate a self-signed certificat: in a terminal, launch the next command line

   `openssl req -x509 -new -utf8 -nodes -out mnemopwdcert -keyout mnemopwdkey`

   where `mnemopwdcert` is the certificat file to share and `mnemopwdkey` is the private key
   file to keep secret (never shared this file)
   
   See https://www.openssl.org/docs/manmaster/apps/req.html for more informations.

2. Test the certificat: in the test/test_certificat directory and in a terminal launch
   the next command line:

   `./server.py -c /path/to/mnemopwdcert -k /path/to/mnemopwdkey`
   
   that start a local server application using the certificat (use `./server.py -h` to get help).

   In a second terminal and in the test/test_certificat directory, launch:

   `./client.py -c /path/to/mnemopwd`

   that start a local client that communicate with the previous local server (use `./client.p -h` to get help).

   If the certificat is accepted, the client shows it and the message `The certificat seems compatible`.

3. Move certificat file and key file in a secure directory on server (I recommend the mnemopwddata
directory created by the server application the first start you launch it).

4. The certificat file (`mnemopwdcert`) must be copy on each client computer to share it.

5. Finally, indicate to the server the `-c` and `-k` options and to the client the `-c` option or 
modify configuration files (by default `~/.mnemopwds` for server and `~/mnemopwdc` for client).
For example (stop server if it is already started):

   `./serverctl --start -c /path/to/mnemopwdcert -k /path/to/mnemopwdkey`


CHANGELOG.rst
Changelog
=========


v1.2.1 (2018-07-10)
-------------------

Changes
~~~~~~~
- New process for generating the session number. [Thierry Lemeunier]
- UML model updated. [thethythy]


v1.2.0 (2017-07-23)
-------------------

New
~~~
- Add clear and secure importation / exportation functionalities.
  [thethythy]
- Add a FileChooserWindow standard dialog box. [thethythy]
- Add a ListBox widget. [thethythy]
- Add the possibility to use colours in client interface (#6)
  [thethythy]

Changes
~~~~~~~
- UML model updated. [thethythy]
- Better UX using the vertical scrollbar widget. [thethythy]
- Replace and delete SearchPanelResult by ListBox widget in
  SearchWindow. [thethythy]
- Better cursor handler in editors. [Thierry Lemeunier]
- Now the configuration file is updated if necessary when a new version
  of MnemoPwd is installed. [Thierry Lemeunier]

Fix
~~~
- Fix a bug in the vertical scroll bar widget. [thethythy]
- Bug correction when LAN is not conform with RFC 1918. [thethythy]


v1.1.0 (2017-03-28)
-------------------
- Add a limitation of number of login and user account creation per hour
  (#2) [thethythy]
- Add IP auto-configuration of the client (enhancement #3) [thethythy]
- ULM model updated. [thethythy]
- Add a scrolling bar in the research result panel (issue #4 fixed)
  [thethythy]
- Add a 3D effect to dialog windows (enhancement #5) [thethythy]


v1.0.0 (2017-02-22)
-------------------
- Add a change log file generated with gitchangelog
  (https://github.com/vaab/gitchangelog) [thethythy]
- UI and UX enhancements. [thethythy]

  - A widget button is 'inversed' only when it obtains the focus
  - Each menu has a border
  - Shortcuts are now in lowercase
- Add the number of blocks loaded in the progress bar. [thethythy]
- Bug #17 fixed (client disconnection when too many blocks are received)
  [thethythy]
- Issue #16 fixed (links between edition window and search window)
  [thethythy]
- Issue #15 fixed (search panel refreshing bug) [thethythy]


v1.0.0rc3 (2016-12-16)
----------------------
- UML model updated (add LabelBox class in the class diagram) [Thierry
  Lemeunier]
- Add the LabelBox widget (issues #13 and #14 fixed) [Thierry Lemeunier]
- Issue #12 fixed (InputBox widget bug) [Thierry Lemeunier]
- Issue #11 fixed (impossible to login from two different OS) [Thierry
  Lemeunier]
- Issue #11 fixed (impossible to login from two different OS) [Thierry
  Lemeunier]


v1.0.0rc2 (2016-11-26)
----------------------
- Update file list to take into account in the fingerprint. [Thierry
  Lemeunier]
- Issues #10 #9 #8 fixed. [Thierry Lemeunier]
- README file updated. [Thierry Lemeunier]
- Add some directories to reject in .gitignore. [Thierry Lemeunier]


v1.0.0rc1 (2016-11-05)
----------------------
- Update README.rst. [Thierry Lemeunier]
- Update README.rst. [Thierry Lemeunier]
- Modification of README file syntax. [Thierry Lemeunier]
- Release candidat v1.0.0rc1. [Thierry Lemeunier]
- Use last version of pyelliptic (fork) [Thierry Lemeunier]
- Make source code compliant much as possible with PEP 8. [Thierry
  Lemeunier]
- Fix a bug when a SIB is truncated by the network. [Thierry Lemeunier]
- UML model updated. [Thierry Lemeunier]
- Now the result panel is updated after an edition, a creation or a
  deletion of a block. [Thierry Lemeunier]
- Add the scrolling of components in the result panel. [Thierry
  Lemeunier]
- Fix a bug in the exception handler of the protocol and another bug in
  user account creation. [Thierry Lemeunier]
- The UI layer use internally SIBs to minimize clear confidential
  information in memory. [Thierry Lemeunier]
- Now connection is correctly closed when user quit application.
  [Thierry Lemeunier]
- Now error messages of the server contain no indication of the source
  of the error. [Thierry Lemeunier]
- Had an error message when the server address is already use. [Thierry
  Lemeunier]
- Add the date of starting and duration in the status message of the
  server. [Thierry Lemeunier]
- Control the terminal size with the number of fields of a block type.
  [Thierry Lemeunier]
- Add screen lock in the edition window and in the search window.
  [Thierry Lemeunier]
- Add a salt value in the password digest computation. [Thierry
  Lemeunier]
- Fix a bug in the configuration at the first launching. [Thierry
  Lemeunier]
- Add an automatic or manual configurable screen lock. [Thierry
  Lemeunier]
- Implementation of the state S32 (client side) [Thierry Lemeunier]
- Minor modification about connection error handling. [Thierry
  Lemeunier]
- User account creation/deletion processes are operational. [Thierry
  Lemeunier]
- Add a window to create a new user window. [Thierry Lemeunier]
- Modification of the menu. [Thierry Lemeunier]
- Search window + bug corrections. [Thierry Lemeunier]
- Implementation of the state S37 (client side) [Thierry Lemeunier]
- Now I use PyCharm IDE for developping. [Thierry Lemeunier]
- Now the notification in the protocolar states is done in a separate
  thread. [Thierry Lemeunier]
- Fix a bug in case there is no result. [Thierry Lemeunier]
- Add a task execution loop inside the main asyncio loop. [Thierry
  Lemeunier]

  - Commands coming from UI layer are registered like tasks
  - Each task is enqueued and then executed one by one
  - The task loop execute each task and wait for the task's end
  - Protocol states are executed in a thread one by one (using a reentrant lock)
- Implementation of the state S34 (client side) [Thierry Lemeunier]
- Add some logs in protocolar states of the server. [Thierry Lemeunier]
- Implementation of the state S36 (client side) [Thierry Lemeunier]
- Implementation of the state S35 (client side) [Thierry Lemeunier]
- Add the block edition window. [Thierry Lemeunier]
- Rationalization of the component hierarchy and add redraw method.
  [thethythy]
- Add block type and the create menu. [thethythy]

  - Block types are loaded from a json file during configuration at starting

mnemopwd/client/ClientUI.py
# -*- coding: utf-8 -*-

# Copyright (c) 2016, Thierry Lemeunier <thierry at lemeunier dot net>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification,
# are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this 
# list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from threading import Thread
#import locale
import curses
from curses.textpad import Textbox

from client.util.funcutils import Observer
from client.util.Configuration import Configuration

"""
Standard curses user interface.
"""

class Component():
    """Abstract component"""
    
    def __init__(self, wparent, y, x):
        self.parent = wparent
        self._cur_y = 0
        self._cur_x = 0
    
    @property
    def cursor_y(self):
        return self._cur_y
    
    @cursor_y.setter
    def cursor_y(self, value):
        self._cur_y = value
    
    @property
    def cursor_x(self):
        return self._cur_x
    
    @cursor_x.setter
    def cursor_x(self, value):
        self._cur_x = value
    
    def isEditable(self):
        """Return False by default (not editable)"""
        return False
    
    def focusOn(self):
        """This component obtains the focus"""
        pass
        
    def focusOff(self):
        """This component losts the focus"""
        pass

class InputBox(Component):
    """A simple text editor with a border line"""
    
    def __init__(self, wparent, h, w, y, x):
        """Create a input text box"""
        Component.__init__(self, wparent, y, x)
        self.panel = wparent.derwin(h, w, y, x)
        self.panel.border()
        self.editorbox = self.panel.derwin(1, w - 4, 1, 2)
        self.editor = Textbox(self.editorbox)
        
    def isEditable(self):
        """This component is editable"""
        return True
        
    def focusOn(self):
        """See mother class"""
        self.editorbox.addstr(self.cursor_y, self.cursor_x, '_', curses.A_BLINK)
        self.editorbox.move(self.cursor_y, self.cursor_x)
        self.editorbox.refresh()
        
    def focusOff(self):
        """See mother class"""
        self.editorbox.move(self.cursor_y, self.cursor_x)
        self.editorbox.clrtoeol()
        self.editorbox.refresh()

class ButtonBox(Component):
    """A simple button text box"""
    
    def __init__(self, wparent, y, x, label):
        Component.__init__(self, wparent, y, x)
        self.label = '[' + label + ']'
        self.parent.addstr(y, x, self.label, curses.A_REVERSE)
        self._cur_y = y
        self._cur_x = x
        
    def focusOn(self):
        """See mother class"""
        self.parent.addstr(self.cursor_y, self.cursor_x, self.label, curses.A_BLINK | curses.A_REVERSE)
        self.parent.refresh()
        
    def focusOff(self):
        """See mother class"""
        self.parent.addstr(self.cursor_y, self.cursor_x, self.label, curses.A_REVERSE)
        self.parent.refresh()

class LoginWindow():
    """
    The login window: get the login/password user credentials
    """
    
    def __init__(self):
        """Create the window"""
        size_y = 14
        size_x = 60
        
        self.loginscr = curses.newwin(size_y, size_x,
                                      int(curses.LINES / 2) - int(size_y / 2),
                                      int(curses.COLS / 2) - int(size_x / 2))
        self.loginscr.border()
        self.loginscr.addstr(1, 2, "Connection window")
        self.loginscr.hline(2,1, curses.ACS_HLINE, size_x - 2)
        self.loginscr.addstr(5, 2, "Login")
        self.loginscr.addstr(8, 2, "Password")
        
        self.logineditor = InputBox(self.loginscr, 3, size_x - 15, 5 - 1, 12)
        self.passeditor = InputBox(self.loginscr, 3, size_x - 15, 8 - 1, 12) 
        
        self.connectButton = ButtonBox(self.loginscr, 11, 7, "Connect")
        self.clearButton = ButtonBox(self.loginscr, 11, 27, "Clear")
        self.quitButton = ButtonBox(self.loginscr, 11, 47, "Quit")
        
        self.items = [self.logineditor, self.passeditor, self.connectButton, self.clearButton, self.quitButton]
        
        self.loginscr.keypad(True)
        self.loginscr.refresh()
    
    def start(self):
        """Start interaction loop of the window"""
        curses.nonl()
        
        self.items[0].focusOn()
        index = 0
        
        while True:
            c = self.loginscr.getch()
            
            if c in [curses.KEY_DOWN, curses.ascii.TAB] :
                self.items[index].focusOff()
                index = (index + 1) % len(self.items)
                self.items[index].focusOn()
                
            elif c in [curses.KEY_UP]:
                self.items[index].focusOff()
                index = (index - 1) % len(self.items)
                self.items[index].focusOn()
                
            elif c in [curses.KEY_LEFT]:
                if (self.items[index].isEditable()):
                    pass
                else:
                    curses.ungetch(curses.KEY_UP)
                    
            elif c in [curses.KEY_RIGHT]:
                if (self.items[index].isEditable()):
                    pass
                else:
                    curses.ungetch(curses.KEY_DOWN)
                    
            elif c in [curses.ascii.CR]:
                if (self.items[index].isEditable()):
                    curses.ungetch(curses.KEY_DOWN)
                else:
                    pass
            
            elif c == ord('q'):
                curses.nl()
                self.loginscr.clear()
                self.loginscr.refresh()
                break
                #return self.logineditor.result(), self.passeditor.result()
                
            else:
                self.loginscr.addstr(curses.unctrl(c) + b' ')
        
        return False, False

class ClientUI(Thread, Observer):

    """
    The default client user interface (UI).
    
    Attribut(s):
    - facade: the facade of the domain layer
    #- encoding: the locale encoding
    - stdscr: the main window of the UI
    - menuscr: the menu window at the top line of the UI
    - statscr: the status window at the bottom line of the UI
    
    Method(s):
    - stop: stop the interaction loop
    - run: start the interaction loop
    - update: update the UI (implementation of the Observer class method)
    """

    def __init__(self, facade):
        """Create an initialize UI"""
        Thread.__init__(self)
        Observer.__init__(self)
        
        self.facade = facade # Store the facade of the domain layer
        
        #locale.setlocale(locale.LC_ALL, '') # Set locale setting
        #self.encoding = locale.getpreferredencoding() # Get locale encoding
        
        # Main window
        self.stdscr = curses.initscr()
        self.stdscr.keypad(True)
        curses.noecho() 
        try: curses.curs_set(0)
        except: pass
        
        # Menu window
        self.menuscr = curses.newwin(2, curses.COLS, 0, 0)
        self.menuscr.hline(1, 0, curses.ACS_HLINE, curses.COLS)
        message = "MnemoPwd Client v" + Configuration.version
        self.menuscr.addstr(0, curses.COLS - len(message) - 1, message)
        self.menuscr.addch(0, curses.COLS - len(message) - 3, curses.ACS_VLINE)
        self.menuscr.addch(1, curses.COLS - len(message) - 3 , curses.ACS_BTEE)
    
        self.menuscr.addstr(0, 0, "S", curses.A_UNDERLINE)
        self.menuscr.addstr(0, 1, "earch")
        self.menuscr.addstr(0, 7, "N", curses.A_UNDERLINE)
        self.menuscr.addstr(0, 8, "ew")
        self.menuscr.addstr(0, 11, "D", curses.A_UNDERLINE)
        self.menuscr.addstr(0, 12, "isconnect")
        self.menuscr.addstr(0, 22, "Q", curses.A_UNDERLINE)
        self.menuscr.addstr(0, 23, "uit")
        self.menuscr.refresh()
        
        # Status window
        self.statscr = curses.newwin(2, curses.COLS, curses.LINES - 2, 0)
        self.statscr.hline(0, 0, curses.ACS_HLINE, curses.COLS)
        self.statscr.attrset(curses.A_DIM)
        self.statscr.refresh()

    def stop(self):
        """Stop UI and return to normal interaction"""
        curses.nocbreak()
        self.stdscr.keypad(False)
        try: curses.curs_set(2)
        except: pass
        curses.echo()
        curses.endwin()

    def run(self):
        """Start the loop interaction"""
        
        # Get login/password
        login, passwd = LoginWindow().start()
        #if (login != False):
            #self.facade.setCredentials(login, passwd)
            #login = passwd = "                            " 

        # Interaction loop
        while True:
            c = self.statscr.getch()
            if c == ord('q'):
                self.facade.stop() # Close the domain layer
                break  # Exit the interaction loop

    def update(self, property, value):
        """Implementation of the method of the class Observer."""
        if property == "connection.state":
            self._update_status(value)
            
    def _update_status(self, value):
        """Update the status window content"""
        currenty, currentx = curses.getsyx() # Save current cursor position
        self.statscr.move(1, 0)
        self.statscr.clrtoeol()
        self.statscr.addstr(value + ' ' + str(currenty) + ' ' + str(currentx))
        #self.statscr.addstr(value)
        self.statscr.refresh()
        curses.setsyx(currenty, currentx)   # Set cursor position to saved position

